clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -110,113 +113,47 @@ private boolean isTransitionValid(State source, State target) {
     }
 
     public TransactionManager(String transactionalId, int transactionTimeoutMs) {
-        pidAndEpoch = new PidAndEpoch(NO_PRODUCER_ID, NO_PRODUCER_EPOCH);
-        sequenceNumbers = new HashMap<>();
+        this.pidAndEpoch = new PidAndEpoch(NO_PRODUCER_ID, NO_PRODUCER_EPOCH);
+        this.sequenceNumbers = new HashMap<>();
         this.transactionalId = transactionalId;
         this.transactionTimeoutMs = transactionTimeoutMs;
-        this.pendingTransactionalRequests = new PriorityQueue<>(10, new Comparator<TransactionalRequest>() {
-            @Override
-            public int compare(TransactionalRequest o1, TransactionalRequest o2) {
-                return Integer.compare(o1.priority().priority(), o2.priority.priority());
-            }
-        });
         this.transactionCoordinator = null;
         this.consumerGroupCoordinator = null;
         this.newPartitionsToBeAddedToTransaction = new HashSet<>();
         this.pendingPartitionsToBeAddedToTransaction = new HashSet<>();
         this.partitionsInTransaction = new HashSet<>();
         this.pendingTxnOffsetCommits = new HashMap<>();
+        this.pendingRequests = new PriorityQueue<>(10, new Comparator<TxnRequestHandler>() {
+            @Override
+            public int compare(TxnRequestHandler o1, TxnRequestHandler o2) {
+                return Integer.compare(o1.priority().priority, o2.priority().priority);
+            }
+        });
     }
 
     public TransactionManager() {
         this("", 0);
     }
 
-    static class TransactionalRequest {
-        private enum Priority {
-            FIND_COORDINATOR(0),
-            INIT_PRODUCER_ID(1),
-            ADD_PARTITIONS_OR_OFFSETS(2),
-            END_TXN(4);
+    // We use the priority to determine the order in which requests need to be sent out. For instance, if we have
+    // a pending FindCoordinator request, that must always go first. Next, If we need a PID, that must go second.
+    // The endTxn request must always go last.
+    private enum Priority {
+        FIND_COORDINATOR(0),
+        INIT_PRODUCER_ID(1),
+        ADD_PARTITIONS_OR_OFFSETS(2),
+        END_TXN(4);
```

Comment:
That's just a mistake. Initially I had separate priorities for `ADD_PARTITIONS` and `ADD_OFFSETS`, but forgot to adjust the `END_TXN` priority after combining them.